src/java/org/apache/cassandra/sensors/ActiveRequestSensors.java

```diff
@@ -0,0 +1,138 @@
+ * entry point to create and modify sensors. More specifically:
+ * <ul>
+ *     <li>Create a new sensor associated to the request/response via {@link #registerSensor(Context, Type)}.</li>
+ *     <li>Increment the sensor value for the request/response via {@link #incrementSensor(Context, Type, double)}.</li>
+ *     <li>Sync this request/response sensor value to the {@link SensorsRegistry} via {@link #syncAllSensors()}.</li>
+ * </ul>
+ * Sensor values related to a given request/response are isolated from other sensors, and the "same" sensor
+ * (for a given context and type) registered to different requests/responses will have a different value: in other words,
+ * there is no automatic synchronization or coordination across sensor values belonging to different
+ * {@link RequestSensors} objects, hence {@link #syncAllSensors()} MUST be invoked to propagate the sensors values
+ * at a global level to the {@link SensorsRegistry}.
+ */
+public class ActiveRequestSensors implements RequestSensors
+{
+    private final Supplier<SensorsRegistry> sensorsRegistry;
+    private final ConcurrentMap<Pair<Context, Type>, Sensor> sensors = new ConcurrentHashMap<>();
+    private final ConcurrentMap<Sensor, Double> latestSyncedValuePerSensor = new ConcurrentHashMap<>();
+    private final ReadWriteLock updateLock = new ReentrantReadWriteLock();
+
+    ActiveRequestSensors()
+    {
+        this(() -> SensorsRegistry.instance);
+    }
+
+    @VisibleForTesting
+    ActiveRequestSensors(Supplier<SensorsRegistry> sensorsRegistry)
+    {
+        this.sensorsRegistry = sensorsRegistry;
+    }
+
+    public void registerSensor(Context context, Type type)
+    {
+        sensors.putIfAbsent(Pair.create(context, type), new Sensor(context, type));
+    }
+
+    public Optional<Sensor> getSensor(Context context, Type type)
+    {
+        return Optional.ofNullable(sensors.get(Pair.create(context, type)));
+    }
+
+    public Set<Sensor> getSensors(Type type)
+    {
+        return sensors.values().stream().filter(s -> s.getType() == type).collect(Collectors.toSet());
+    }
+
+    public void incrementSensor(Context context, Type type, double value)
+    {
+        updateLock.readLock().lock();
+        try
+        {
+            Optional.ofNullable(sensors.get(Pair.create(context, type))).ifPresent(s -> s.increment(value));
+        }
+        finally
+        {
+            updateLock.readLock().unlock();
+        }
+    }
+
+    public void syncAllSensors()
+    {
+        updateLock.writeLock().lock();
+        try
+        {
+            sensors.values().forEach(sensor -> {
+                double current = latestSyncedValuePerSensor.getOrDefault(sensor, 0d);
+                double update = sensor.getValue() - current;
+                if (update == 0d)
+                    return;
+
+                latestSyncedValuePerSensor.put(sensor, sensor.getValue());
```

Comment:
I think more correct name is `previousSyncedValuePerSensor` because the actual update in registry happens after

Reply:
I guess it depends how you see it :) It's the latest before the current update, or the previous after the current update :) I'd keep the current name, also it's unrelated to this PR.